File: python/day3.py
from collections import Counter

from utils import Solution, read_text, test


def get_test(filename):
    return list(map(int, read_text(filename).split(",")))


class Day3(Solution):
    @staticmethod
    def part1():
        numbers = get_test(test(3, 1))
        return sum(set(numbers))

    @staticmethod
    def part2():
        numbers = get_test(test(3, 2))
        return sum(sorted(set(numbers))[:20])

    @staticmethod
    def part3():
        numbers = get_test(test(3, 3))
        return max(Counter(numbers).values())


if __name__ == "__main__":
    Day3.run()

File: rust/src/day_3.rs
use crate::utils::Solution;
use std::collections::{BTreeSet, HashMap, HashSet};
use std::fmt::Display;

pub struct Day3;

impl Solution for Day3 {
    const DAY: u32 = 3;

    fn part1(input: &str) -> impl Display {
        input
            .split(",")
            .map(|s| s.parse::<i32>().unwrap())
            .collect::<HashSet<_>>()
            .iter()
            .copied()
            .sum::<i32>()
    }

    fn part2(input: &str) -> impl Display {
        input
            .split(",")
            .map(|s| s.parse::<i32>().unwrap())
            .collect::<BTreeSet<_>>()
            .iter()
            .take(20)
            .copied()
            .sum::<i32>()
    }

    fn part3(input: &str) -> impl Display {
        input
            .split(",")
            .map(|s| s.parse::<i32>().unwrap())
            .fold(HashMap::new(), |mut acc, num| {
                *acc.entry(num).or_insert(0) += 1;
                acc
            })
            .values()
            .copied()
            .max()
            .unwrap()
    }
}
